library/mps/reader.c

```diff
@@ -0,0 +1,626 @@
+
+    TRACE( trace_comment, "Final state: (end=commit,fo,avail) = (%u,%u,%u)",
+           (unsigned) end, (unsigned) fo, (unsigned) aa );
+    RETURN( 0 );
+}
+
+int mbedtls_reader_reclaim( mbedtls_reader *rd, size_t *paused )
+{
+    unsigned char *frag, *acc;
+    size_t pause, commit;
+    size_t al, fo, fl;
+    TRACE_INIT( "reader_reclaim" );
+
+    if( paused != NULL )
+        *paused = 0;
+
+    frag = rd->frag;
+    if( frag == NULL )
+        RETURN( MBEDTLS_ERR_READER_UNEXPECTED_OPERATION );
+
+    acc    = rd->acc;
+    pause  = rd->pause;
+    commit = rd->commit;
+    fl     = rd->frag_len;
+
+    if( acc == NULL )
+        fo = 0;
+    else
+        fo = rd->acc_share.frag_offset;
+
+    if( pause == 0 )
+    {
+        TRACE( trace_comment, "No unsatisfied read-request has been logged." );
+        /* Check if there's data left to be consumed. */
+        if( commit < fo || commit - fo < fl )
+        {
+            TRACE( trace_comment, "There is data left to be consumed." );
+            rd->end = commit;
+            RETURN( MBEDTLS_ERR_READER_DATA_LEFT );
+        }
+        TRACE( trace_comment, "The fragment has been completely processed and committed." );
+    }
+    else
+    {
+        size_t frag_backup_offset;
+        size_t frag_backup_len;
+        TRACE( trace_comment, "There has been an unsatisfied read-request with %u bytes overhead.",
+               (unsigned) pause );
+
+        if( acc == NULL )
+        {
+            TRACE( trace_comment, "No accumulator present" );
+            RETURN( MBEDTLS_ERR_READER_NEED_ACCUMULATOR );
+        }
+        al = rd->acc_len;
+
+        /* Check if the upper layer has already fetched
+         * and committed the contents of the accumulator. */
+        if( commit < fo )
+        {
+            /* No, accumulator is still being processed. */
+            TRACE( trace_comment, "Still processing data from the accumulator" );
+            if( fo + fl < fo              ||
+                fo + fl + pause < fo + fl ||
+                al < fo + fl + pause )
+            {
+                rd->end = commit;
+                rd->pause = 0;
+                TRACE( trace_error, "The accumulator is too small to handle the backup." );
+                TRACE( trace_error, "* Remaining size: %u", (unsigned) al );
+                TRACE( trace_error, "* Needed: %u (%u + %u + %u)",
+                       (unsigned) ( fo + fl + pause ),
+                       (unsigned) fo, (unsigned) fl, (unsigned) pause );
+                RETURN( MBEDTLS_ERR_READER_ACCUMULATOR_TOO_SMALL );
+            }
+            frag_backup_offset = 0;
+            frag_backup_len = fl;
+        }
+        else
+        {
+            /* Yes, the accumulator is already processed. */
+            TRACE( trace_comment, "The accumulator has already been processed" );
+            frag_backup_offset = commit;
+            frag_backup_len = fl - commit;
+
+            if( frag_backup_len + pause < pause ||
```

Comment:
same comment as above - pause is used twice

Reply:
As above, this is an unsigned integer overflow check.